exercises/exercise-13/database.ts

```diff
@@ -1,13 +1,85 @@
-export class Database<T> {
+import * as fs from 'fs';
+
+type JsonScalar = boolean | number | string;
+
+export type FieldOp = |
+    {$eq: JsonScalar} |
+    {$gt: JsonScalar} |
+    {$lt: JsonScalar} |
+    {$in: JsonScalar[]};
+
+export type Query<T extends object> = |
+    {$and: Query<T>[]} |
+    {$or: Query<T>[]} |
+    {$text: string} |
+    (
+        {[field in QueryableKeys<T>]?: FieldOp}
+    );
+
+function matchOp(op: FieldOp, v: JsonScalar) {
+    if ('$eq' in op) {
+        return v === op['$eq'];
+    } else if ('$gt' in op) {
+        return v > op['$gt'];
+    } else if ('$lt' in op) {
+        return v < op['$lt'];
+    } else if ('$in' in op) {
+        return op['$in'].includes(v);
+    }
+    throw new Error(`Unrecognized op: ${op}`);
+}
+
+type IndexedRecord<T extends object> = T & {
+    $index: {[word: string]: true};
+};
+
+type Unionize<T extends object> = {[k in keyof T]: {k: k, v: T[k]}}[keyof T];
+type QueryableKeys<T extends object> = Extract<Unionize<T>, {v: JsonScalar}>['k'];
```

Comment:
This enforces that you only run queries on fields whose values are `JsonScalar` types. Maybe useful as an advanced extension checked with `@ts-expect-error`.